File: java/TowerOfHanoi.java
import java.util.LinkedList;

/**
 * Implementácia známeho problému presunu veže, zloženej z krúžkov rôznej veľkosti, z jedného miesta na druhé.
 * 
 * https://en.wikipedia.org/wiki/Tower_of_Hanoi
 */
public class TowerOfHanoi {
    
    private final int size;

    private LinkedList<Integer> firstStick = new LinkedList<>();
    private LinkedList<Integer> secondStick = new LinkedList<>();
    private LinkedList<Integer> thirdStick = new LinkedList<>();

    private int stepCount = 0;

    public TowerOfHanoi(int size) {
        this.size = size;
        for (int i = 1; i <= size; i++) {
            firstStick.addFirst(i);
        }
    }

    public void solve() {
        System.out.println("Starting ...");
        prinntSticks();
        while (thirdStick.size() < this.size) {
            // Algoritmus sa líši poradím krokov podľa toho, či máme párny alebo nepárny počet krúžkov.
            if (this.size % 2 == 1) {
                if (makeOneMove(firstStick, secondStick, 1, 2)) {
                    break;
                }
                if (makeOneMove(firstStick, thirdStick, 1, 3)) {
                    break;
                }
            } else {
                if (makeOneMove(firstStick, thirdStick, 1, 3)) {
                    break;
                }
                if (makeOneMove(firstStick, secondStick, 1, 2)) {
                    break;
                }
            }
            if (makeOneMove(secondStick, thirdStick, 2, 3)) {
                break;
            }
        }
        System.out.println("Finished in " + this.stepCount + " steps");
        prinntSticks();
    }

    private void prinntSticks() {
        System.out.println("Stick 1: " + this.firstStick);
        System.out.println("Stick 2: " + this.secondStick);
        System.out.println("Stick 3: " + this.thirdStick);
    }

    /**
     * Skúsi presunúť vrchnú položku z veže A na vežu B alebo opačne.
     * Vráti true, ak sú po tomto kroku všetky krúžky presunuté na tretiu vežu, inak false.
     */
    private boolean makeOneMove(
            LinkedList<Integer> stickA, 
            LinkedList<Integer> stickB, 
            int stickNumberA, 
            int stickNumberB) {

        if (moveIfPossible(stickA, stickB)) {
            this.stepCount ++;
            System.out.println(stickNumberA + " -> " + stickNumberB);
        } else if (moveIfPossible(stickB, stickA)) {
            this.stepCount ++;
            System.out.println(stickNumberB + " -> " + stickNumberA);
        }
        if (thirdStick.size() == this.size) {
            return true;
        } 
        return false;
    }

    private boolean moveIfPossible(LinkedList<Integer> from, LinkedList<Integer> to) {
        if (!from.isEmpty() && (to.isEmpty() || to.getLast() > from.getLast())) {
            to.addLast(from.removeLast());
            return true;
        }
        return false;
    }

    public static void main(String[] args) {
        if (args.length != 1) {
            System.err.println("Usage: java TowerOfHanoi size");
            System.exit(1);
        }
        int size = Integer.parseInt(args[0], 10);
        TowerOfHanoi toverOfHanoi = new TowerOfHanoi(size);
        toverOfHanoi.solve();
    }
}

File: python/tower_of_hanoi.py
#
# Implementácia známeho problému presunu veže, zloženej z krúžkov rôznej veľkosti, z jedného miesta na druhé.
# https://en.wikipedia.org/wiki/Tower_of_Hanoi
#

from typing import List

diskCount = int(input('Enter number of disks:'))

tower1 = [ *range(diskCount, 0, -1) ]
tower2 = []
tower3 = []

stepCount = 0

def swap_disks(firstTower: List[int], secondTower: List[int], firstName, secondName):
    global stepCount
    if len(firstTower) == 0 and len(secondTower) == 0:
        return
    stepCount += 1
    if len(firstTower) == 0 or len(secondTower) and firstTower[-1] > secondTower[-1]:
        print(secondName, '->', firstName)
        firstTower.append(secondTower.pop())
        return
    if len(secondTower) == 0 or len(firstTower) and secondTower[-1] > firstTower[-1]:
        print(firstName, '->', secondName)
        secondTower.append(firstTower.pop())

print(tower1, tower2, tower3)

while len(tower3) < diskCount:
    if diskCount % 2 == 1:
        swap_disks(tower1, tower2, 1, 2)
        if len(tower3) == diskCount:
            break
        swap_disks(tower1, tower3, 1, 3)
        
        if len(tower3) == diskCount:
            break
    else:
        swap_disks(tower1, tower3, 1, 3)
        if len(tower3) == diskCount:
            break
        swap_disks(tower1, tower2, 1, 2)
        if len(tower3) == diskCount:
            break

    swap_disks(tower2, tower3, 2, 3)

print(tower1, tower2, tower3, 'in', stepCount, 'steps')